README.md
# node-ses
==========

A simple and reliable Node.js mail for sending mail through Amazon SES.

## Benefits

 * Does only one thing and does it well. Only the [SendEmail](http://docs.aws.amazon.com/ses/latest/APIReference/API_SendEmail.html) and [SendRawEmail](http://docs.aws.amazon.com/ses/latest/APIReference/API_SendRawEmail.html) API methods are implemented.
 * Good error handling:
   * Only "2xx" and "3xx" responses from Amazon are considered successful.
   * Returned error objects are the _Error_ elements of [Amazon's error responses](http://docs.aws.amazon.com/ses/latest/DeveloperGuide/query-interface-responses.html) with _Type_, _Code_, _Message_ etc.
   * Support for the `debug` module is included if [debugging](#debugging) is needed.
 * Tested and reliable. Includes test suite. Sending email to SES since 2012.

## Synopsis

Start by creating a client object, and then call either the `sendEmail` or `sendRawEmail` method
depending on your needs.

```javascript
var ses = require('node-ses')
  , client = ses.createClient({key: 'key', secret: 'secret'});

// Give SES the details and let it construct the message for you.
client.sendEmail({
   to: 'aaron.heckmann+github@gmail.com'
 , from: 'somewhereOverTheR@inbow.com'
 , cc: 'theWickedWitch@nerds.net'
 , bcc: ['canAlsoBe@nArray.com', 'forrealz@.org']
 , subject: 'greetings'
 , message: 'your <b>message</b> goes here'
 , altText: 'plain text'
}, function (err, data, res) {
 // ...
});

// ... or build a message from scratch yourself and send it.
client.sendRawEmail({
 , from: 'somewhereOverTheR@inbow.com'
 , rawMessage: rawMessage
}, function (err, data, res) {
 // ...
});
```

## Installation

`npm install node-ses`

The module has one primary export:

## createClient()

You'll probably only be using this method. It takes an options object with the following properties:

    `key` -  your AWS SES key. Defaults to checking `process.env.AWS_ACCESS_KEY_ID` and `process.env.AWS_ACCESS_KEY`
    `secret` - your AWS SES secret. Defaults to `process.env.AWS_SECRET_ACCESS_KEY` and `process.env.AWS_SECRET_KEY`
    `amazon` - [optional] the amazon end-point uri. defaults to `https://email.us-east-1.amazonaws.com`

Not all AWS regions support SES. Check [SES region support](http://docs.aws.amazon.com/ses/latest/DeveloperGuide/regions.html) to be sure the region you are in is supported.


```js
var ses = require('node-ses')
  , client = ses.createClient({ key: 'key', secret: 'secret' });
```

## client.sendEmail(options, function (err, data, res))

Composes an email message based on input data, and then immediately queues the message for sending.

There are several important points to know about SendEmail:

 * You can only send email from verified email addresses and domains; otherwise, you will get an "Email address not verified" error. If your account is still in the Amazon SES sandbox, you must also verify every recipient email address except for the recipients provided by the Amazon SES mailbox simulator. For more information, go to the [Amazon SES Developer Guide](http://docs.aws.amazon.com/ses/latest/DeveloperGuide/verify-addresses-and-domains.html).
 * The total size of the message cannot exceed 10 MB. This includes any attachments that are part of the message.
 * Amazon SES has a limit on the total number of recipients per message. The combined number of To:, CC: and BCC: email addresses cannot exceed 50. If you need to send an email message to a larger audience, you can divide your recipient list into groups of 50 or fewer, and then call Amazon SES repeatedly to send the message to each group.
 * For every message that you send, the total number of recipients (To:, CC: and BCC:) is counted against your sending quota - the maximum number of emails you can send in a 24-hour period. For information about your sending quota, go to the [Amazon SES Developer Guide](http://docs.aws.amazon.com/ses/latest/DeveloperGuide/manage-sending-limits.html).


`sendEmail` receives an options object with the following properties:

    `from` - email address from which to send (required)
    `subject` - string (required). Must be encoded as UTF-8
    `message` - can be html (required). Must be encoded as UTF-8.
    `altText` - plain text version of message. Must be encoded as UTF-8.
    `to` - email address or array of addresses
    `cc` - email address or array of addresses
    `bcc` - email address or array of addresses
    `replyTo` - email address
    `configurationSet` - SES configuration set name
    `messageTags` - SES message tags: array of name/value objects, e.g. { name: xid, value: 1 }

At least one of `to`, `cc` or `bcc` is required.

Optional properties (overrides the values set in `createClient`):

    `key` - AWS key
    `secret` - AWS secret
    `amazon` - AWS end point. Defaults to `https://email.us-east-1.amazonaws.com`

The `sendEmail` method transports your message to the AWS SES service. If AWS
returns an HTTP status code that's less than `200` or greater than or equal to
400, we will callback with an `err` object that is a direct tranalation of the _Error_ element of the AWS error response.

See [Error Handling](#error-handling) section below for details on the structure of returned errors.

Check for errors returned since a 400 status is not uncommon.

The `data` returned in the callback is an object containing the parsed AWS response.

See the [SES API Response](http://docs.aws.amazon.com/ses/latest/DeveloperGuide/query-interface-responses.html) docs for details.

The `res` returned by the callback represents the HTTP response to calling the SES REST API as the [request](https://www.npmjs.org/package/request) module returns it.

*The sendEmail method also be  provided in all lowercase as `sendemail` for backwards compatibility.*

## client.sendRawEmail(options, function (err, data, res))

Sends an email message, with header and content specified by the client. The SendRawEmail action is useful for sending multipart MIME emails. The raw text of the message must comply with Internet email standards; otherwise, the message cannot be sent.

There are several important points to know about SendRawEmail:

 * You can only send email from verified email addresses and domains; otherwise, you will get an "Email address not verified" error. If your account is still in the Amazon SES sandbox, you must also verify every recipient email address except for the recipients provided by the Amazon SES mailbox simulator. For more information, go to the [Amazon SES Developer Guide](http://docs.aws.amazon.com/ses/latest/DeveloperGuide/verify-addresses-and-domains.html).
 * The total size of the message cannot exceed 10 MB. This includes any attachments that are part of the message.
 * Amazon SES has a limit on the total number of recipients per message. The combined number of To:, CC: and BCC: email addresses cannot exceed 50. If you need to send an email message to a larger audience, you can divide your recipient list into groups of 50 or fewer, and then call Amazon SES repeatedly to send the message to each group.
 * The To:, CC:, and BCC: headers in the raw message can contain a group list. Note that each recipient in a group list counts towards the 50-recipient limit.
For every message that you send, the total number of recipients (To:, CC: and BCC:) is counted against your sending quota - the maximum number of emails you can send in a 24-hour period. For information about your sending quota, go to the [Amazon SES Developer Guide](http://docs.aws.amazon.com/ses/latest/DeveloperGuide/manage-sending-limits.html).


`sendRawEmail` receives an options object with the following properties:

    `from` - email address from which to send (required)
    `rawMessage` - the raw text of the message which includes a header and a body (required)

Within the raw text of the message, the following must be observed:

* The `rawMessage` value must contain a header and a body, separated by a blank line.
* All required header fields must be present.
* Each part of a multipart MIME message must be formatted properly.
* MIME content types must be among those supported by AWS SES. For more information, see the [SES Developer Guide](http://docs.aws.amazon.com/ses/latest/DeveloperGuide/mime-types.html).
* The `rawMessage` content must be base64-encoded, if MIME requires it.

The `sendRawEmail` method transports your message to the AWS SES service. If Amazon
returns an HTTP status code that's less than `200` or greater than or equal to
400, we will callback with an `err` object that is a direct translation of the _Error_ element of aws error response.

See [Error Handling](#error-handling) section below for details on the structure of returned errors.

### Example

```js
var CRLF = '\r\n'
  , ses = require('node-ses')
  , client = ses.createClient({ key: 'key', secret: 'secret' })
  , rawMessage = [
    'From: "Someone" <someone@example.com>',
    'To: "Someone Else" <other@example.com>',
    'Subject: greetings',
    'Content-Type: multipart/mixed;',
    '    boundary="_003_97DCB304C5294779BEBCFC8357FCC4D2"',
    'MIME-Version: 1.0',
    '',
    '--_003_97DCB304C5294779BEBCFC8357FCC4D2',
    'Content-Type: text/plain; charset="us-ascii"',
    'Content-Transfer-Encoding: quoted-printable',
    'Hi brozeph,',
    '',
    'I have attached a code file for you.',
    '',
    'Cheers.',
    '',
    '--_003_97DCB304C5294779BEBCFC8357FCC4D2',
    'Content-Type: text/plain; name="code.txt"',
    'Content-Description: code.txt',
    'Content-Disposition: attachment; filename="code.txt"; size=4;',
    '    creation-date="Mon, 03 Aug 2015 11:39:39 GMT";',
    '    modification-date="Mon, 03 Aug 2015 11:39:39 GMT"',
    'Content-Transfer-Encoding: base64',
    '',
    'ZWNobyBoZWxsbyB3b3JsZAo=',
    '',
    '--_003_97DCB304C5294779BEBCFC8357FCC4D2',
    ''
  ].join(CRLF);

client.sendRawEmail({
 , from: 'someone@example.com'
 , rawMessage: rawMessage
}, function (err, data, res) {
 // ...
});
```

Check for errors returned since a 400 status is not uncommon.

The `data` returned in the callback is an object containing the parsed Amazon json response.

See the [SES API Response](http://docs.aws.amazon.com/ses/latest/DeveloperGuide/query-interface-responses.html) docs for details.

The `res` returned by the callback represents the HTTP response to calling the SES REST API as the [request](https://www.npmjs.org/package/request) module returns it.

<a name="error-handling"></a>
## Error Handling

The errors returned when sending failed are JavaScript objects that correspond to the Error element as seen in [Structure of an Error Response](http://docs.aws.amazon.com/ses/latest/DeveloperGuide/query-interface-responses.html) from the official documentation. The properties in error object we return include:

 * A `Type` element that identifies whether the error was a Receiver, Sender, or NodeSesInternal error
 * A `Code` element that identifies the type of error that occurred
 * A `Message` element that describes the error condition in a human-readable form
 * A `Detail` element that might give additional details about the error or might be empty

An error of Type `NodeSesInternal` is returned in three cases:

 * If the HTTP request to AWS fails so that we don't get a normal response from AWS. The `Code` will be `RequestError` and the `Message` will contain the HTTP request error.
 * If aws error response has invalid schema (Error element is missing), then the `Code` will be set to `XmlError` and the `Message` will contain explanation and the original response.

Example error response:

```json
{
  "Type": "Sender",
  "Code": "ValidationError",
  "Message": "Value null at 'message.subject' failed to satisfy constraint: Member must not be null"
}
```

<a name="debugging"></a>
## Debugging

```bash
# Enable in the shell
DEBUG="node-ses" ./server.js
```

```javascript
// ... or temporarily set in your code before `node-ses` is required.
process.env.DEBUG='node-ses';
```


When debugging, it's useful to inspect the raw HTTP request and response send
to Amazon. These can then checked against Amazon's documentation for the [SendMail](http://docs.aws.amazon.com/ses/latest/APIReference/API_SendEmail.html) API method and the [common errors](http://docs.aws.amazon.com/ses/latest/APIReference/CommonErrors.html) that Amazon might return.

To turn on debugging printed to STDERR, set `DEBUG=node-ses` in the environment before running your script. You can also set `process.env.DEBUG='node-ses';` in your code, before the `node-ses` module is required.

See the [debug module](https://www.npmjs.org/package/debug) docs for more debug output possibilities.

## Running the Tests

Unit tests

 `npm test`

To run the full tests, including actually using the AWS SES REST APIs with your credentials, set the following environment variables:

    # Your SES Key and secret
    NODE_SES_KEY
    NODE_SES_SECRET

    # An email that is both an verified sende that can also receive test emails. Possibly your own email address
    NODE_SES_EMAIL

## See Also

 * [node-ses-any-promise](https://www.npmjs.com/package/node-ses-any-promise) is a fork with a promise-based API.
 * [nodemailer](https://www.npmjs.com/package/nodemailer) has more features, including attachment support. There are many "transport" plugins available for it, including one for SES.

## License

[MIT](https://github.com/aheckmann/node-ses/blob/master/LICENSE)


History.md

3.0.2 / 2020-05-04
==================

Fix botched 3.0.1 release.

3.0.1 / 2020-05-04
===================

Revert broken attempt to switch to switch to JSOn parsing in 3.0.0


3.0.0 / 2020-04-16
==================

## [BREAKING CHANGES]

Internally we switched from requesting XML to requesting JSON from AWS.

While the JavaScript API we provide remains the same, there are places where we
used to return errors with a `Code` property set to `XmlError`. In these cases
the error will be set to `JsonError` instead. If your use depended on this
behavior it will need to be changed to explect the new error type instead.
Switching from XML to JSON to allows us to drop the XML dependencies we had,
reducing exposure to potential XML vulnerablities. Since JSON is faster to
parse than XML, performance may improve as well.

2.2.2 / 2020-04-16
==================

## [BUG FIXES]

 * Fix signature signing for eu-central-1 region (#60)

## [INTERNALS]

 * Bump lodash, aws4, request deps

2.2.1 / 2019-05-22
==================

[INTERNALS]

 * TypeScript types added (#56)

2.2.0 / 2019-02-14 ❤️
======================

[NEW FEATURES]

 * Add support for setting the configuration set during sending (#53)

[INTERNALS]

 * Bumped dependencies
 * Dropped support for Node.js < 6. (Although it may still work)


2.1.0 / 2017-11-22
==================

[NEW FEATURES]

 * Allow setting key and secret from the environment. Helps support AWS Lambda. (#49)

2.0.5 / 2017-03-28
==================

[BUG FIXES]

 * sendEmail 'key' option now properly overrides 'key' option given to create() (#43)

2.0.4 / 2017-02-28
==================

[INTERNALS]

 * Add error checking for another case that could cause exceptions to be thrown when AWS has problems. (#42)


2.0.3 / 2016-12-09
===================

[BUG FIXES]

 * Really release fix from 2.0.2 (@lyuzashi)

2.0.2 / 2016-12-05
===================

[BUG FIXES]

 * Fix syntax issue which could break builds in some cases (@otech47)

2.0.1 / 2016-09-28
===================

No code changes.

[INTERNALS]

 * Link to new node-ses-any-promise package, a fork with promise-based API.


2.0.0 / 2016-08-16
==================

There's a major version bump because the internals changed signficantly as part
of switching to version 4 authentication signatures with AWS. The upgrade
is backwards compatible with 1.2 for normal use.

[NEW FEATURES]

 * Add webpack compatibility (cspeer)

[INTERNALS]

 * Switched from using "version 2" AWS authentication signatures to "version 4" for future-proofing. (markstos)
 * Processing the response from AWS has been moved to it's own function. Automated test coverage for it has been added
   and there's now the potential to override how response handling works. For now the function remains undocumented. (markstos)

[BREAKING CHANGES]

 * 'algorithm' property is no longer exported and the option to change it is
   removed. If you are using an alternate algorithm to sign version 4
   Authentication signatures, a pull request is welcome to restore this
   feature. (markstos)


1.2.0 / 2016-04-07
==================

[DOCUMENTATION]

 * More detail on the structure of errors that might be returned was added.
   See new Error Handling heading in the README (markstos)

[INTERNALS]

 * Internal errrors are now returned in the same format as AWS errors for consistency.
   This changes the structure of the error returned for the edge cases where the HTTP request to AWS failed
   or we failed to parse the XML of the response. (markstos)

1.1.0 / 2015-08-25
==================

[BUG FIXES]

 * Documented default value of Amazon end point was wrong (#21, #25, imatveev, DelvarWorld)

[NEW FEATURES]
 
 * Start supporting sendRawEmail method. (#20, brozeph)

[INTERNALS]

 * Fixed typo in source code (#24, DelvarWorld)

1.0.2 / 2015-07-23
==================

[DOCUMENTATION]

 * Improved debugging documentation

[INTERNALS]
 
 * Bump dependencies
 * Increase test time-out

1.0.1 / 2015-04-15
==================

[BUG FIXES]

 * We now allow the `cc` or `bcc` fields to be used instead of `to`, mirroring the SES API. (#11, Cellule)

[DOCUMENTATION]

 * Explicitly document format for SES endpoints and clarify that not all AWS regions support SES. (#13, rickwaugh1)
 * We now document that we expect input to be UTF-8 encoded.

[INTERNALS]

  * xml2json dependency with replaced with smaller xml2js dependency. [cromestant](https://github.com/cromestant) and  [markstos](https://github.com/markstos) (#15)
  * Bump dependencies

1.0.0 / 2014-11-06
==================

[THINGS THAT MIGHT BREAK YOUR CODE]

  * The format of SES errors has changed from XML to JavaScript objects. This should fit in more naturally
    with Node.js code, but is not backward compatible if you happened to be parsing the errors before.  [robludwig](https://github.com/robludwig)

[DOCUMENTATION]

  * docs were updated to clarify what the `data` and `res` return values are [markstos](https://github.com/markstos) (#6)

0.0.3 / 2012-08-20
==================

  * fixed; fail gracefully when not connected to the network [jessetane](https://github.com/jessetane) (#3)

0.0.2 / 03-14-2012
==================

  * docs
  * add missing dependency - #1

0.0.1 / 03-13-2012
==================

  * initial release


lib/ses.js
// node-ses
'use strict';

var email = require('./email')
  , DEFAULT_API_HOST = 'https://email.us-east-1.amazonaws.com';


/**
 * SESClient
 *
 * `options` is an object with these properties:
 *
 *    key - your AWS SES key
 *    secret - your AWS SES secret
 *    amazon - [optional] the amazon end-point uri. defaults to amazon `https://email.us-east-1.amazonaws.com`
 *
 * Example:
 *
 *     var client = new SESClient({ key: 'key', secret: 'secret' });
 *     client.sendemail({
 *        to: 'aaron.heckmann+github@gmail.com'
 *      , from: 'somewhereOverTheR@inbow.com'
 *      , cc: 'theWickedWitch@nerds.net'
 *      , bcc: ['canAlsoBe@nArray.com', 'forrealz@.org']
 *      , subject: 'greetings'
 *      , message: 'your message goes here'
 *      , altText: 'mmm hmm'
 *    }, function (err) {
 *      // ...
 *    });
 *
 * @param {Object} options
 */
function SESClient (options) {
  options = options || {};
  this.key = options.key;
  this.secret = options.secret;
  this.amazon = options.amazon || exports.amazon;
}


/**
 * Send an email
 *
 * @param {Object} options
 * @param {Function} callback
 */
SESClient.prototype.sendEmail = function (options, callback) {
  options.key = options.key || this.key;
  options.secret = options.secret || this.secret;
  options.amazon = options.amazon || this.amazon;
  options.action = email.actions.SendEmail;

  var message = new email.Email(options);
  message.send(callback);
};

/**
 * Send an email (convenience alias to sendemail)
 *
 * @param {Object} options
 * @param {Function} callback
 **/
SESClient.prototype.sendemail = SESClient.prototype.sendEmail;

/**
 * Send an email (raw)
 *
 * @param {Object} options
 * @param {Function} callback
 */
SESClient.prototype.sendRawEmail = function (options, callback) {
  options.key = options.key || this.key;
  options.secret = options.secret || this.secret;
  options.amazon = options.amazon || this.amazon;
  options.action = email.actions.SendRawEmail;

  var message = new email.Email(options);
  message.send(callback);
};

/**
 * Exports
 **/
exports.createClient = function createClient (options) {
  return new SESClient(options);
};

exports.Email = email.Email;
exports.amazon = DEFAULT_API_HOST;
exports.version = require('../package.json').version;


node-ses.d.ts
import { Response } from "request";

declare module "node-ses" {
  export type Emails = string | string[];
  export interface MessageTag {
    name: string;
    value: any;
  }
  export interface sendEmailOptions {
    from: string;
    subject: string;
    message: string;
    altText?: string;
    to?: Emails;
    cc?: Emails;
    bcc?: Emails;
    replyTo?: string;
    configurationSet?: string;
    messageTags?: MessageTag[];
    key?: string;
    secret?: string;
    amazon?: string;
  }
  export interface sendRawEmailOptions {
    from: string;
    rawMessage: string;
  }
  export interface SendEmailResult {
    MessageId: string;
  }
  export interface ResponseMetadata {
    RequestId: string;
  }
  export interface SendEmailResponse {
    SendEmailResult: SendEmailResult;
    ResponseMetadata: ResponseMetadata;
  }
  export interface SendEmailData {
    SendEmailResponse: SendEmailResponse;
  }
  export interface SendEmailError {
    Type: string;
    Code: string;
    Message: string;
    Detail?: string;
  }
  export type Callback = (
    error: SendEmailError,
    data: SendEmailData,
    response: Response
  ) => any;
  export interface Client {
    sendEmail(options: sendEmailOptions, callback: Callback): void;
    sendemail(options: sendEmailOptions, callback: Callback): void;
    sendRawEmail(options: sendRawEmailOptions, callback: Callback): void;
  }
  export function createClient({
    key,
    secret,
    amazon
  }: {
    key: string;
    secret: string;
    amazon?: string;
  }): Client;
}


lib/email.js
'use strict';

const aws4 = require('aws4');
const debug = require('debug')('node-ses');
const parse = require('url').parse;
const querystring = require('querystring');
const request = require('request');
const xml2js = require('xml2js');
const xmlParser = new xml2js.Parser({explicitArray:false,mergeAttrs:true});

const SEND_EMAIL_ACTION = 'SendEmail';
const SEND_RAW_EMAIL_ACTION = 'SendRawEmail';

/**                                                                        
 * Email constructor.
 *
 * @param {Object} options
 **/
function Email (options) {
  this.action = options.action || SEND_EMAIL_ACTION;
  this.key = options.key;
  this.secret = options.secret;
  this.amazon = options.amazon;
  this.from = options.from;
  this.subject = options.subject;
  this.message = options.message;
  this.altText = options.altText;
  this.rawMessage = options.rawMessage;
  this.configurationSet = options.configurationSet;
  this.messageTags = options.messageTags;
  this.extractRecipient(options, 'to');
  this.extractRecipient(options, 'cc');
  this.extractRecipient(options, 'bcc');
  this.extractRecipient(options, 'replyTo');
}


/**
 * Adds to, cc, and bcc fields to `data`.
 *
 * @param {Object} data
 * @return data
 **/
Email.prototype.addDestination = function (data) {
  this.to.forEach(function (to, i) {
    data['Destination.ToAddresses.member.' + (i + 1)] = to;
  });

  this.cc.forEach(function (to, i) {
    data['Destination.CcAddresses.member.' + (i + 1)] = to;
  });

  this.bcc.forEach(function (to, i) {
    data['Destination.BccAddresses.member.' + (i + 1)] = to;
  });

  return data;
};

/**
 * Adds ConfigurationSetName and Tags to `data`
 *
 * @param {Object} data
 * @return data
 **/
Email.prototype.addSESHeaders = function (data) {
  if (this.configurationSet) {
    data['ConfigurationSetName'] = this.configurationSet;
  }
  if (this.messageTags) {
    this.messageTags.forEach(function (tagSpec, i) {
      data['Tags.member.' + (i + 1) + '.Name'] = tagSpec.name;
      data['Tags.member.' + (i + 1) + '.Value'] = tagSpec.value;
    });
  }
  return data;
};

/**
 * Adds subject, alt text, and message body to `data`.
 *
 * @param {Object} data
 * @return data
 **/
Email.prototype.addMessage = function (data) {
  if (this.subject) {
    data['Message.Subject.Data'] = this.subject;
    data['Message.Subject.Charset'] = 'UTF-8';
  }

  if (this.message) {
    data['Message.Body.Html.Data'] = this.message;
    data['Message.Body.Html.Charset'] = 'UTF-8';
  }

  if (this.altText) {
    data['Message.Body.Text.Data'] = this.altText;
    data['Message.Body.Text.Charset'] = 'UTF-8';
  }

  return data;
};


/**
 * Adds the list of ReplyTos to `data`.
 *
 * @param {Object} data
 * @return data
 **/
Email.prototype.addReplyTo = function (data) {
  this.replyTo.forEach(function(to, i) {
    data['ReplyToAddresses.member.' + (i + 1)] = to;
  });

  return data;
};


/**
 * Prepares param object for the AWS request.
 *
 * @return {Object}
 */
Email.prototype.data = function () {
  var data = {
      Action: this.action
    , AWSAccessKeyId: this.key
    , Source: this.from
  };

  // recipients and reply tos
  data = this.addDestination(data);
  data = this.addReplyTo(data);

  // SES Event Publishing Information
  data = this.addSESHeaders(data);

  // message payload
  if (this.action === SEND_EMAIL_ACTION) {
    data = this.addMessage(data);
  } else if (this.action === SEND_RAW_EMAIL_ACTION) {
    data['RawMessage.Data'] = Buffer.from(this.rawMessage).toString('base64');
  }

  return data;
};


/**
 * Extracts recipients from options.
 *
 * @param {String} prop - either to,cc,bcc,replyTo
 * @param {Object} options
 */
Email.prototype.extractRecipient = function (options, prop) {
  if (options[prop]) {
    this[prop] = Array.isArray(options[prop]) ? options[prop] : [options[prop]];
  } else {
    this[prop] = [];
  }
};


/**
 * Creates required AWS headers.
 *
 * No additional custom headers by default.
 *
 * @return Object
 **/
Email.prototype.headers = function () {
  var headers = {};
  return headers;
};

/**
 * Sends the email.
 *
 * @param {Function} callback
 * @api Public
 **/
Email.prototype.send = function send (callback) {
  const invalid = this.validate();

  if (invalid) {
    return callback(new Error(invalid));
  }

  // Prepare the data and send to it AWS SES REST API

  var data = querystring.stringify(this.data());
  var parsedUrl = parse(this.amazon);
  var headers = this.headers();

  var options = {
      uri: this.amazon
    , host: parsedUrl.hostname
    , headers: headers
    , body: data
    , service: 'ses'
  };

  var credentials = this.key && this.secret && {
    accessKeyId : this.key,
    secretAccessKey : this.secret
  };
  var signedOpts = aws4.sign(options, credentials);

  signedOpts.headers['Connection'] = 'keep-alive';

  debug('posting: %j', signedOpts);

  request(signedOpts, (err, res, data) => this._processResponse(err, res, data, callback));
};

/**
 * Process a response from the 'request' call used to POST email to SES
 * @private
 *
 */
Email.prototype._processResponse = function _processResponse (err, res, data, callback) {
  debug('received: %s %d %j', err, res && res.statusCode || 0, data);

  if (err) {
    return callback({
      Type : "NodeSesInternal",
      Code : "RequestError",
      Message: err,
    }, data, res);
  }

  if (res.statusCode < 200 || res.statusCode >= 400) {
    return xmlParser.parseString(data,function(err,result){
        if(err) {
          return callback({
            Type: "NodeSesInternal",
            Code: "ParseError",
            Message: err
          });
        }

        // Return an error object with keys of Type, Code and Message
        // Reference docs at: http://docs.aws.amazon.com/ses/latest/DeveloperGuide/query-interface-responses.html
        if(result && result.ErrorResponse){
          return callback(result.ErrorResponse.Error)
        }
        // Once during an S3 outage AWS returned valid XML that didn't match the usual error structure.
        // We cover this unlikely case
        else {
          return callback({
            Type : "NodeSesInternal",
            Code : "XmlError",
            Message: result,
          })
        }

    });
  } else {
    return callback(null, data, res);
  }
};

/**
 * Validates the input.
 *
 * @returns {String|Undefined}
 **/
Email.prototype.validate = function () {
  if (this.action === SEND_EMAIL_ACTION) {
    if (!this.to.length && !this.cc.length && !this.bcc.length) {
      return 'To, Cc or Bcc is required';
    }

    if (!(this.subject && this.subject.length)) {
      return 'Subject is required';
    }
  }

  if (this.action === SEND_RAW_EMAIL_ACTION) {
    if (!(this.rawMessage && this.rawMessage.length)) {
      return 'Raw message is required';
    }
  }

  // all actions require the following
  if (!(this.from && this.from.length)) {
    return 'From is required';
  }
};

/**
 * Exports
 **/
exports.Email = Email;

exports.actions = {
  SendEmail : SEND_EMAIL_ACTION,
  SendRawEmail : SEND_RAW_EMAIL_ACTION
};
